tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -1654,11 +1669,26 @@ void HWCommandQueue::enqueue_program(Program& program, bool blocking) {
                 false);
         }
     }
+#ifdef DEBUG
+    if (getenv("TT_METAL_VALIDATE_PROGRAM_BINARIES") != nullptr) {
+        TT_FATAL(!this->manager.get_bypass_mode(), "Tracing cannot be used while validating program binaries");
+        for (int buffer_idx = 0; buffer_idx < program.program_transfer_info.kernel_bins.size(); buffer_idx++) {
+            const auto& buffer = program.kg_buffers[buffer_idx];
+            std::vector<uint32_t> read_data(buffer->page_size() * buffer->num_pages() / sizeof(uint32_t));
+            // Use Slow Dispatch read instead?
+            this->enqueue_read_buffer(*buffer, read_data.data(), true);
```

Comment:
You probably want to use SD, since you might perturb the CQ state with FD. Although maybe skip the check for non-mmio chips? This readback will take forever with SD..

Reply:
Right, but even with SD we perturb the state because we need to block before reading, which is why I just left it as a FD read which also works well for non-mmio.

Though SD read would perturb cq state less.

Reply:
I think this is fine - we'll either be looking at program corruption or CQ state, probably not both at once